Service Worker: Implement FetchEvent.replacesClientId

gterzian opened the pull request:
#### f65cd19397ece138300da2494b03314c848ec290
<pre>
Service Worker: Implement FetchEvent.replacesClientId

<a href="https://bugs.webkit.org/show_bug.cgi?id=226638">https://bugs.webkit.org/show_bug.cgi?id=226638</a>

Reviewed by NOBODY (OOPS!).

This changes adds a replacesClientId to fetch events triggered by a navigation, leaving it to an empty string otherwise.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5bf10e2f86c8bf5c3c192da23aa34dc4e158684

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23336 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1450 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24459 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21558 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23607 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23892 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23577 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1008 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20209 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26103 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21096 "Passed tests") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27254 "An unexpected error occured. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21279 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21358 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25123 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/808 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18554 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/765 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1221 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1067 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-event.https.html:
```diff
@@ -147,6 +147,28 @@
         });
   }, 'Service Worker responds to fetch event with the correct resulting client id');
+  
+  promise_test(t => {
```
Is this a test resync from upstream wpt ? If not, you will need to upstream those changes.

gterzian replied:
This is a new test that I wrote. Ok, I will look into how to upstream it.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/fetch-event-test-worker.js:
```diff
@@ -47,6 +47,16 @@ function handleResultingClientId(event) {
   event.respondWith(new Response(body));
 }
 
+function handleReplacesClientId(event) {
```
Ditto.

cdumez reviewed (approved):
LGTM. @youennf please take a look too if possible.

youennf commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2192,6 +2192,7 @@ void DocumentLoader::loadMainResource(ResourceRequest&& request)
         // The main navigation load will trigger the registration of the client.
         if (m_resultingClientId)
             scriptExecutionContextIdentifierToLoaderMap().remove(m_resultingClientId);
+            mainResourceLoadOptions.replacesClientIdentifier = m_resultingClientId.object();
```
Bad indent or missing { }.
Also I am not sure this is the right way to get the resultingClientId, given a FrameLoader has several document loaders.
It might be best to get the id from the frame document instead if there is such a document.

gterzian replied:
> Bad indent or missing { }.

Ah ok, because it is now two lines of code in the conditional block...

> It might be best to get the id from the frame document instead if there is such a document.

Ok, I will look further into this. Also, the spec refers to this being the id of the active document env settings object, so one could look up which one is active for the frame?

youennf commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-event.https.html:
```diff
@@ -147,6 +147,28 @@
   }, 'Service Worker responds to fetch event with the correct resulting client id');
+  
+  promise_test(t => {
+      const page_url = 'resources/simple.html?replacesClientId';
```
This is a same origin navigation.
I wonder what we should do for cross origin navigations. It might be good to write a test, maybe for cross origin redirection as well.
Are we expected to expose cross origin client IDs given the service worker cannot do a lot of things with it?

gterzian replied:
>  It might be good to write a test, maybe for cross origin redirection as well.

Ok I can add one. 

> Are we expected to expose cross origin client IDs given the service worker cannot do a lot of things with it?

I think so yes, since I cannot find anything to the contrary in the spec. cc @annevk 

The worker could use the id to clear some data related to the client that is about to go away?

annevk replied:
Hmm. A `Client` object is static right and represents what the service worker intercepts? How could it represent a cross-origin environment?

I guess that's possible with `openWindow()`? That seems wrong-ish.

gterzian replied:
FYI had a look at https://bugs.webkit.org/attachment.cgi?id=449475&action=diff, but it seems the previously implemented `targetClientId` was only implemented as a default empty string, it wasn't actually set yet. So no guidance there. 

I think `openWindow()` should show an empty `replaclesClientId` because it loads the url in a newly created top-level bc, so the one being replaced would be "about:blank". 

It seems the cross-origin case would rather be about a client navigating an existing iframe across origin? If the resulting client id is exposed in that case, then the one being replaced could be exposed as well?

annevk replied:
When A embeds A1 and A1 is navigated to B, A's service worker would only service A1, not B. B would have its own service worker. I don't think we'd expose a client of B to A? (If we do, I think that would be a bug, probably a security bug.)

youennf replied:
In case of A1 -> B cross-origin navigation, we would start exposing A1 id to B service worker.
I am not sure there is a security issue here but I don't really understand what B service worker would do with A1 id in case of cross-origin navigation.
Are we already exposing cross origin env IDs elsewhere? If this is new, why should we start doing it?

annevk replied:
We don't really expose any information cross-origin currently, not even URL. E.g., `location.href` only has a cross-origin setter. So that definitely seems like a bug.

gterzian replied:
> A1 -> B cross-origin navigation, we would start exposing A1 id to B service worker

How does `resultingClientId` currently deals with the mirror of that example(the future id of B being exposed in the fetch event to A's sw, with A embedding A1)?

gterzian replied:
> A1 -> B cross-origin navigation, we would start exposing A1 id to B service worker

Also, if A1 is embedded as a frame in A, and they are same-origin, with A being a client to a sw. If A1 is navigated to a cross-origin destination B(through JS manipulated it in the context of A), then a fetch event will be delivered to the sw of A, and not to what will be the sw of B after the navigation completes, right? If that is correct, then the `replacesClientId` will not be cross-origin, it will be that of A1 prior to the navigation, and it will be delivered to the same origin sw of A. 

If A1 is cross origin to begin with, then it cannot be navigated by code in A(right?). If it can, the risk is rather that both the `replacesClientId`, and the `resultingClientId`, will both be cross-origin yet exposed to the sw of A. 

By the way, is it to correct to say that the only use-case of `replacesClientId` is the navigation of an embedded frame, since in other cases `clientId` will always equal `replacesClientId`?

youennf replied:
Navigation loads are different from sub resource loads as they trigger matching service worker.
See https://w3c.github.io/ServiceWorker/#on-fetch-request-algorithm step 14.6 (vs step 15 for sub resources).
Navigation loads are always handled by same origin service workers, hence why resultingClientId is also always same origin.
In case of cross origin navigation redirections, a new resultingClientId will be generated. 

> By the way, is it to correct to say that the only use-case of replacesClientId is the navigation of an embedded frame, since in other cases clientId will always equal replacesClientId?

Top level frame navigations may expose replacesClientId.

gterzian replied:
> Navigation loads are different from sub resource loads as they trigger matching service worker.

Thanks for the info! 

In that case, it appears to me that the `replacesClientId` will always be that of a client whose origin matches the worker receiving the fetch event. Except maybe in the cross-origin redirect case.

youennf replied:
>  it appears to me that the replacesClientId will always be that of a client whose origin matches the worker receiving the fetch event. Except maybe in the cross-origin redirect case.

Can you clarify? My understanding is that `replacesClientId` would be cross origin in case of any cross origin navigation, except if the spec is filtering somehow exposure of the env id doing the navigation.

gterzian replied:
Ok, sorry I was confused about the matching algorithm, though it would match on the worker of the origin where the navigation request originates, while it is in fact matching on the "target" origin. 

There is even an assertion in step 8.2 [the matching algorithm](https://w3c.github.io/ServiceWorker/#match-service-worker-registration) that the origin of the url of the navigation and the resulting matching scope are equal(added [here](https://github.com/w3c/ServiceWorker/issues/1118)). 

So, it feels as if [Step 24.3.9 of Handle Fetch](https://w3c.github.io/ServiceWorker/#handle-fetch) should include a same origin check and initialize `replacesClientId` to the empty string if it fails?

Could one use the request's client origin, and compare it with the request's url?

youennf replied:
Yes, that should do the trick spec wise, it would be good to discuss this with the service worker WG.
The thing to cover well in terms of WPT tests will be redirections, ideally both iframes and top level navigations.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/fetch-event-test-worker.js:
```diff
@@ -47,6 +47,16 @@ function handleResultingClientId(event) {
+function handleReplacesClientId(event) {
+  var body;
+  if (event.replacesClientId !== "") {
+    body = 'Replaces Client ID Found: ' + event.replacesClientId;
```
For same origin clients, we should be able to get the Client from its id and do some checks.

gterzian replied:
Ok, will look into it.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-event.https.html:
```diff
@@ -147,6 +147,28 @@
+      const expected_found = 'Replaces Client ID Found';
+      const expected_not_found = 'Replaces Client ID Not Found';
+      return with_iframe(page_url)
+        .then(function(frame) {
```
You can use await if you like it

gterzian replied:
Ok, thanks.

gterzian committed:
Service Worker: Implement FetchEvent.replacesClientId

https://bugs.webkit.org/show_bug.cgi?id=226638

Reviewed by NOBODY (OOPS!).

This changes adds a replacesClientId to fetch events triggered by a navigation, leaving it to an empty string otherwise.
